### PR TITLE
feat: Use websocket to continuously surface runtime log into the UI

### LINKED ISSUE #6511: Support activity and bot state inspection in the embedded Web Chat control
- [x] Capture message / trace / event activities (P0)
- [x] Capture network activity (P0)
- [x] Show selected item in inspector (P0)
- [ ] Fix layout so that Web Chat panel **does not** overlap with inspector pane (P2)
- [x] Expose runtime messages in "Output" tab (P0)
- [ ]~~Add different inspector experience for LUIS traces (P1)~~
- [ ] Selected inspected state in log -- link log item to item shown in inspector pane (P1)
- [ ] On initial load, select most recent activity (P1)

### PR DESCRIPTION
**Pull request by srinaath**
## Description

Instead of polling constantly, using a websocket to deliver the stdout and stderr from a started runtime.

Output Window padding fixes.

## Task Item

Fixes #6511 #6750

### CODE REVIEW

File: Composer/packages/client/src/components/BotRuntimeController/ErrorCallout.tsx [hunk 1]
```diff
@@ -81,6 +78,8 @@ fieldsWhiteList.set('code', { visible: false, name: 'code' });
 fieldsWhiteList.set('signal', { visible: false, name: 'signal' });
 fieldsWhiteList.set('stderr', { visible: false, name: 'stderr' });
 
+const DEFAULT_FONT_SETTINGS = getDefaultFontSettings();
```
Comment (hatpick) [inline Composer/packages/client/src/components/BotRuntimeController/ErrorCallout.tsx:81]: why is this using the "default" font settings instead font settings?

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx [hunk 2]
```diff
@@ -21,15 +33,73 @@ const createMarkup = (txt: string) => {
...
+  useEffect(() => {
+    const setupLogConnection = async () => {
+      try {
+        const runtimeStreamUrl = await httpClient.get(`/publish/runtimeLogUrl/${projectId}`);
+
+        runtimeTrafficChannel.current = new WebSocket(runtimeStreamUrl.data);
+
+        if (runtimeTrafficChannel.current) {
+          runtimeTrafficChannel.current.onmessage = (event) => {
+            try {
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx:59]: why do you need this try within the outer try, the catch block is doing the same thing for both

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx [hunk 3]
```diff
@@ -38,33 +108,31 @@ export const RuntimeOutputLog: React.FC<{ projectId: string }> = ({ projectId })
...
+            wordBreak: 'break-all',
+            whiteSpace: 'pre-wrap',
+            lineHeight: '20px',
+          }}
+          // eslint-disable-next-line react/no-danger
+          dangerouslySetInnerHTML={createMarkup(runtimeData.standardOutput)}
+          data-testid="runtime-standard-output"
+        />
+      )}
+      {botBuildErrors && <ErrorCallout error={botBuildErrors} />}
```
Comment (hatpick) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx:134]: does botBuildErrors ever has no value?

File: extensions/localPublish/src/WebSocketServer.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (hatpick) [inline extensions/localPublish/src/WebSocketServer.ts:1]: nit: javascript convention for classes is camelCase naming for the file (in oppose to Java and c# which is PascalCase)

File: extensions/localPublish/src/WebSocketServer.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import portfinder from 'portfinder';
+import express, { Request, Response } from 'express';
+import { Server as WSServer } from 'ws';
+
+interface WebSocket {
+  close(): void;
+  send(data: string, cb?: (err?: Error) => void): void;
+}
+
+export class WebSocketServer {
```
Comment (hatpick) [inline extensions/localPublish/src/WebSocketServer.ts:15]: I'd give this a better name to reflect on what the purpose of the web socket server is.

File: extensions/localPublish/src/WebSocketServer.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+          this.servers[projectId] = wsServer;
+        }
+      });
+      this.port = port;
+      return this.port;
+    }
+  }
+
+  public static sendRuntimeLogToSubscribers(projectId: string, standardOutput: string, standardError: string): void {
+    this.sockets[projectId]?.send(
```
Comment (hatpick) [inline extensions/localPublish/src/WebSocketServer.ts:76]: dont you wanna know if the projectId is invalid and someone is trying to send a message on its socket?

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx (hunk 3 above)
Comment (srinaath) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx:134]: It has a value only if there are build errors. For ex LUIS key invalid, publish to LUIS failed.

File: extensions/localPublish/src/WebSocketServer.ts (hunk 5 above)
Comment (srinaath) [reply extensions/localPublish/src/WebSocketServer.ts:15]: Yep. Changing it to runtimeLogServer

File: extensions/localPublish/src/WebSocketServer.ts (hunk 6 above)
Comment (srinaath) [reply extensions/localPublish/src/WebSocketServer.ts:76]: If there is a subscriber for the projectID it should exist. If there is no subscriber there is actually no client (composer UI) for me to send this runtime log for a project.

File: Composer/packages/client/src/components/BotRuntimeController/ErrorCallout.tsx (hunk 1 above)
Comment (srinaath) [reply Composer/packages/client/src/components/BotRuntimeController/ErrorCallout.tsx]: Removed it. Its using the font of the parent.

File: extensions/localPublish/package.json [hunk 7]
```diff
@@ -13,20 +13,25 @@
     "@botframework-composer/types": "file:../../Composer/packages/types",
     "adm-zip": "^0.4.14",
     "archiver": "^5.0.2",
+    "express": "^4.17.1",
+    "express-serve-static-core": "^0.1.1",
```
Comment (tonyanziano) [inline extensions/localPublish/package.json:17]: is this `...-serve-static-core` package needed? I'm not seeing it used anywhere in the diff

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx [hunk 8]
```diff
@@ -21,15 +33,69 @@ const createMarkup = (txt: string) => {
...
+
+    if (!runtimeTrafficChannel.current) {
+      setupLogConnection();
+    }
+
+    return () => {
+      runtimeTrafficChannel.current?.close();
+      runtimeTrafficChannel.current = null;
+    };
+  }, []);
```
Comment (tonyanziano) [inline Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx:98]: should this function list `projectId` as a dependency? What happens if you open a second project? Will it have only opened the first, initial socket for project A even though you are now using project B?

File: extensions/localPublish/src/WebSocketServer.ts (hunk 6 above)
Comment (hatpick) [reply extensions/localPublish/src/WebSocketServer.ts]: I do understand the logic of how it works, but the ```?.``` accessor is showing that there might be a projectId value that leads to an undefined socket. That's a bad state that shouldn't happen but when it happens you definitely wanna know about it to fix it perhaps.

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx (hunk 3 above)
Comment (hatpick) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx:130]: the default value in the recoil state shows an object, meaning it always has a value, when you reset it falls back to that default value right?

File: Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx (hunk 8 above)
Comment (srinaath) [reply Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/RuntimeOutputLog/RuntimeOutputLog.tsx:98]: So, i would have added if the component destroy wasn't called on project change. Essentially, when we change project the component is destroyed and the socket as well.

File: extensions/localPublish/package.json (hunk 7 above)
Comment (srinaath) [reply extensions/localPublish/package.json]: Nice catch. I had ported it from server but things work fine without it.

**Review by tonyanziano — APPROVED** [review]

**Pull request merged**